linera-views/src/views/bucket_queue_view.rs

```diff
@@ -0,0 +1,656 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(with_metrics)]
+use std::sync::LazyLock;
+use std::{
+    collections::{vec_deque::IterMut, VecDeque},
+    fmt::Debug,
+};
+
+use async_trait::async_trait;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+#[cfg(with_metrics)]
+use {
+    linera_base::prometheus_util::{self, MeasureLatency},
+    prometheus::HistogramVec,
+};
+
+use crate::{
+    batch::Batch,
+    common::{from_bytes_option, from_bytes_option_or_default, HasherOutput, MIN_VIEW_TAG},
+    context::Context,
+    hashable_wrapper::WrappedHashableContainerView,
+    views::{ClonableView, HashableView, Hasher, View, ViewError},
+};
+
+#[cfg(with_metrics)]
+/// The runtime of hash computation
+static BUCKET_QUEUE_VIEW_HASH_RUNTIME: LazyLock<HistogramVec> = LazyLock::new(|| {
+    prometheus_util::register_histogram_vec(
+        "bucket_queue_view_hash_runtime",
+        "BucketQueueView hash runtime",
+        &[],
+        Some(vec![
+            0.001, 0.003, 0.01, 0.03, 0.1, 0.2, 0.3, 0.4, 0.5, 0.75, 1.0, 2.0, 5.0,
+        ]),
+    )
+    .expect("Histogram can be created")
+});
+
+/// Key tags to create the sub-keys of a BucketQueueView on top of the base key.
+#[repr(u8)]
+enum KeyTag {
+    /// Prefix for the front of the view
+    Front = MIN_VIEW_TAG,
+    /// Prefix for the storing of the stored-indices information
+    Store,
+    /// Prefix for the indices of the log.
+    Index,
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+struct StoredIndices {
+    indices: Vec<(usize, usize)>,
+    position: usize,
+}
+
+impl StoredIndices {
+    fn is_empty(&self) -> bool {
+        self.indices.len() == 0
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Cursor {
+    position: Option<(usize, usize)>,
+}
+
+impl Cursor {
+    pub fn new(stored_indices: &StoredIndices) -> Self {
+        if stored_indices.indices.is_empty() {
+            Cursor { position: None }
+        } else {
+            Cursor {
+                position: Some((0, stored_indices.position)),
+            }
+        }
+    }
+
+    pub fn is_incrementable(&self) -> bool {
+        self.position.is_some()
+    }
+}
+
+/// A view that supports a FIFO queue for values of type `T`.
+pub struct BucketQueueView<C, T, const N: usize> {
```

Comment:
Is `N` the bucket size? Does this really need to be a const generic rather than a field? Won't this generate new code for each value of `N` we use?

Reply:
Yes, `N` is the bucket size. And yes, code will be generated for each instance used in the code. I do not know the details of the code generation in Rust, but I know better how things go in C++. It ought to be similar.

The recommendation in C++ is that there is no downside to having many specific type instances. In any case, the planned usage are about 5-10 instances, so I do not think that it will be a problem.

I do not see how we could have a bucket size that is a field. That size should turn out from the `Self::load` function. If the view is empty, then we cannot get the size.